File: Secure-Multi-Agent-Systems-with-A2A-Protocol-and-AsgardeoSecure-Multi-Agent-Systems-with-A2A-Protocol-and-Asgardeo/auth/agent_auth.py
"""
Asgardeo 3-Step Agent Authentication.
Used by Orchestrator to get its own actor_token for delegation.
"""

import logging
import httpx
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)


class AgentAuthService:
    """
    3-step agent authentication per Asgardeo docs:
    1. POST /oauth2/authorize with response_mode=direct
    2. POST /oauth2/authn with agent credentials
    3. POST /oauth2/token to get actor_token
    """
    
    def __init__(
        self,
        client_id: str,
        client_secret: str,
        agent_id: str,
        agent_secret: str,
        org_name: str = "a2abasic",
        redirect_uri: str = "http://localhost:8000/callback"
    ):
        self.client_id = client_id
        self.client_secret = client_secret
        self.agent_id = agent_id
        self.agent_secret = agent_secret
        self.org_name = org_name
        self.redirect_uri = redirect_uri
        
        # Asgardeo endpoints
        base_url = f"https://api.asgardeo.io/t/{org_name}"
        self.authorize_endpoint = f"{base_url}/oauth2/authorize"
        self.authn_endpoint = f"{base_url}/oauth2/authn"
        self.token_endpoint = f"{base_url}/oauth2/token"
        
        # Cached actor token
        self._actor_token: Optional[str] = None
        
        logger.info(f"AgentAuthService initialized for client: {client_id}")
    
    async def _exchange_code_for_token(self, client: httpx.AsyncClient, code: str, code_verifier: str = None) -> Optional[str]:
        """Exchange code for token (step 3 of the flow)."""
        logger.info("Exchanging code for actor_token...")
        
        token_data = {
            'grant_type': 'authorization_code',
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'code': code,
            'redirect_uri': self.redirect_uri
        }
        
        # Add code_verifier for PKCE if provided
        if code_verifier:
            token_data['code_verifier'] = code_verifier
        
        token_response = await client.post(
            self.token_endpoint,
            headers={'Content-Type': 'application/x-www-form-urlencoded'},
            data=token_data
        )
        
        if token_response.status_code != 200:
            logger.error(f"Token exchange failed: {token_response.status_code}")
            logger.error(f"Response: {token_response.text}")
            return None
        
        token_data = token_response.json()
        self._actor_token = token_data.get('access_token')
        
        if self._actor_token:
            logger.info(f"✅ Actor token obtained: {self._actor_token[:30]}...")
            return self._actor_token
        else:
            logger.error("No access_token in response")
            return None
    
    async def get_actor_token(self, scopes: list = None) -> Optional[str]:
        """
        Get actor_token via 3-step flow.
        Caches the token for reuse.
        """
        if self._actor_token:
            return self._actor_token
        
        if not scopes:
            scopes = ['openid', 'profile']
        
        try:
            async with httpx.AsyncClient() as client:
                # Step 1: Authorize (get flowId)
                logger.info("Step 1: Initiating agent authorization...")
                
                # Create Basic auth header
                import base64
                import secrets
                import hashlib
                
                credentials = f"{self.client_id}:{self.client_secret}"
                basic_auth = base64.b64encode(credentials.encode()).decode()
                
                # Generate PKCE
                code_verifier = secrets.token_urlsafe(64)[:64]
                digest = hashlib.sha256(code_verifier.encode()).digest()
                code_challenge = base64.urlsafe_b64encode(digest).rstrip(b'=').decode()
                
                auth_request_data = {
                    'response_type': 'code',
                    'redirect_uri': self.redirect_uri,
                    'scope': ' '.join(scopes),
                    'response_mode': 'direct',
                    'client_id': self.client_id,
                    'code_challenge': code_challenge,
                    'code_challenge_method': 'S256'
                }
                
                logger.info(f"Authorize request to: {self.authorize_endpoint}")
                logger.info(f"With client_id: {self.client_id}")
                
                auth_response = await client.post(
                    self.authorize_endpoint,
                    headers={
                        'Accept': 'application/json',
                        'Content-Type': 'application/x-www-form-urlencoded',
                        'Authorization': f'Basic {basic_auth}'
                    },
                    data=auth_request_data
                )
                
                if auth_response.status_code != 200:
                    logger.error(f"Step 1 failed: {auth_response.status_code}")
                    logger.error(f"Response: {auth_response.text}")
                    return None
                
                # Debug: log the full response
                response_text = auth_response.text
                logger.info(f"Step 1 response: {response_text[:500]}...")
                
                auth_result = auth_response.json()
                flow_id = auth_result.get('flowId')
                
                # Check if we got a code directly (app-native auth might skip interactive flow)
                if not flow_id:
                    code = auth_result.get('code')
                    if code:
                        logger.info(f"Got code directly (no interactive flow): {code[:20]}...")
                        # Skip step 2, go directly to step 3
                        return await self._exchange_code_for_token(client, code, code_verifier)
                    
                    logger.error(f"No flowId in response. Keys: {list(auth_result.keys())}")
                    return None
                
                # Get authenticator ID - FIX: use auth_result not auth_data
                authenticators = auth_result.get('nextStep', {}).get('authenticators', [])
                if not authenticators:
                    logger.error("No authenticators found")
                    return None
                
                authenticator_id = authenticators[0].get('authenticatorId')
                logger.info(f"Step 1 complete. FlowId: {flow_id[:20]}...")
                
                # Step 2: Authenticate with agent credentials
                logger.info("Step 2: Authenticating agent...")
                
                authn_response = await client.post(
                    self.authn_endpoint,
                    headers={'Content-Type': 'application/json'},
                    json={
                        'flowId': flow_id,
                        'selectedAuthenticator': {
                            'authenticatorId': authenticator_id,
                            'params': {
                                'username': self.agent_id,
                                'password': self.agent_secret
                            }
                        }
                    }
                )
                
                if authn_response.status_code != 200:
                    logger.error(f"Step 2 failed: {authn_response.status_code}")
                    logger.error(f"Response: {authn_response.text}")
                    return None
                
                # Debug: log full response
                authn_text = authn_response.text
                logger.info(f"Step 2 response: {authn_text[:500]}...")
                
                authn_data = authn_response.json()
                
                # Check if auth is complete
                flow_status = authn_data.get('flowStatus')
                if flow_status == 'SUCCESS_COMPLETED':
                    # Check for authData containing code
                    auth_data = authn_data.get('authData', {})
                    code = auth_data.get('code') or authn_data.get('code')
                    if code:
                        logger.info(f"Step 2 complete. Code: {code[:20]}...")
                    else:
                        logger.error(f"flowStatus=SUCCESS_COMPLETED but no code. authData keys: {list(auth_data.keys())}")
                        return None
                else:
                    code = authn_data.get('code')
                    if not code:
                        logger.error(f"No code in authn response. flowStatus={flow_status}, keys: {list(authn_data.keys())}")
                        return None
                    logger.info(f"Step 2 complete. Code: {code[:20]}...")
                
                # Step 3: Exchange code for token
                logger.info("Step 3: Exchanging code for actor_token...")
                
                token_response = await client.post(
                    self.token_endpoint,
                    headers={'Content-Type': 'application/x-www-form-urlencoded'},
                    data={
                        'grant_type': 'authorization_code',
                        'client_id': self.client_id,
                        'client_secret': self.client_secret,
                        'code': code,
                        'redirect_uri': self.redirect_uri,
                        'code_verifier': code_verifier
                    }
                )
                
                if token_response.status_code != 200:
                    logger.error(f"Step 3 failed: {token_response.status_code}")
                    logger.error(f"Response: {token_response.text}")
                    return None
                
                token_data = token_response.json()
                self._actor_token = token_data.get('access_token')
                
                if self._actor_token:
                    logger.info(f"✅ Actor token obtained: {self._actor_token[:30]}...")
                    return self._actor_token
                else:
                    logger.error("No access_token in response")
                    return None
                    
        except Exception as e:
            logger.error(f"Agent auth error: {e}")
            import traceback
            logger.error(traceback.format_exc())
            return None

File: agent-auth-playground/lib/agentAuth.ts
/**
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import 'server-only';

import { randomBytes, createHash } from 'crypto';
import https from 'node:https';
import http from 'node:http';
import { AuthErrorStage, parseOAuthErrorBody } from './authTrace';

function isLocalHost(urlStr: string): boolean {
  try {
    const { hostname } = new URL(urlStr);
    return hostname === 'localhost' || hostname === '127.0.0.1' || hostname === '::1';
  } catch {
    return false;
  }
}

// Bypasses TLS certificate verification for local IS instances (self-signed certs).
function insecureFetch(url: string, init?: RequestInit): Promise<Response> {
  return new Promise<Response>((resolve, reject) => {
    const parsed = new URL(url);
    const useHttps = parsed.protocol === 'https:';
    const transport = useHttps ? https : http;

    const reqOptions: https.RequestOptions = {
      hostname: parsed.hostname,
      port: Number(parsed.port) || (useHttps ? 443 : 80),
      path: parsed.pathname + parsed.search,
      method: (init?.method ?? 'GET').toUpperCase(),
      headers: init?.headers as Record<string, string> | undefined,
      rejectUnauthorized: false,
    };

    const req = transport.request(reqOptions, (res) => {
      const chunks: Buffer[] = [];
      res.on('data', (c: Buffer) => chunks.push(c));
      res.on('end', () => {
        const headers: Record<string, string> = {};
        const raw = res.rawHeaders ?? [];
        for (let i = 0; i + 1 < raw.length; i += 2) {
          headers[raw[i].toLowerCase()] = raw[i + 1];
        }
        resolve(new Response(Buffer.concat(chunks), { status: res.statusCode ?? 200, headers }));
      });
      res.on('error', reject);
    });

    req.on('error', reject);

    const body = init?.body;
    if (body != null) {
      req.write(body instanceof URLSearchParams ? body.toString() : body as string);
    }
    req.end();
  });
}

export function localAwareFetch(url: string, init?: RequestInit): Promise<Response> {
  return isLocalHost(url) ? insecureFetch(url, init) : fetch(url, init);
}

export interface AgentAuthConfig {
  baseUrl: string;
  clientId: string;
  redirectUri: string;
  agentId: string;
  agentSecret: string;
  scope?: string;
}

interface PKCEPair {
  codeVerifier: string;
  codeChallenge: string;
}

interface AuthorizeResult {
  flowId: string;
  authenticatorId: string;
}

interface AuthorizeResponse {
  flowId?: string;
  nextStep?: {
    authenticators?: Array<{ authenticatorId: string }>;
  };
}

interface AuthnResponse {
  authData?: { code?: string };
  code?: string;
  flowStatus?: string;
  failureReason?: string;
  error?: string;
  error_description?: string;
}

interface TokenResponse {
  access_token: string;
  token_type: string;
  expires_in?: number;
  scope?: string;
  [key: string]: unknown;
}

export interface AuthFlowErrorInit {
  stage: AuthErrorStage;
  statusCode?: number;
  errorCode?: string;
  errorDescription?: string;
  url?: string;
  body?: string;
  message?: string;
}

export class AuthFlowError extends Error {
  stage: AuthErrorStage;
  statusCode?: number;
  errorCode?: string;
  errorDescription?: string;
  url?: string;
  body?: string;

  constructor(init: AuthFlowErrorInit) {
    const headline =
      init.message ||
      init.errorDescription ||
      init.errorCode ||
      `${init.stage} failed${init.statusCode ? ` (HTTP ${init.statusCode})` : ''}`;
    super(headline);
    this.name = 'AuthFlowError';
    this.stage = init.stage;
    this.statusCode = init.statusCode;
    this.errorCode = init.errorCode;
    this.errorDescription = init.errorDescription;
    this.url = init.url;
    this.body = init.body;
  }
}

function generatePKCE(): PKCEPair {
  const codeVerifier = randomBytes(48).toString('base64url');
  const codeChallenge = createHash('sha256').update(codeVerifier).digest('base64url');
  return { codeVerifier, codeChallenge };
}

async function readErrorBody(res: Response): Promise<string> {
  try {
    return await res.text();
  } catch {
    return '';
  }
}

async function initiateAuthorize(
  baseUrl: string,
  clientId: string,
  redirectUri: string,
  scope: string,
  codeChallenge: string
): Promise<AuthorizeResult> {
  const url = `${baseUrl}/oauth2/authorize`;
  const body = new URLSearchParams({
    client_id: clientId,
    response_type: 'code',
    redirect_uri: redirectUri,
    scope,
    response_mode: 'direct',
    code_challenge: codeChallenge,
    code_challenge_method: 'S256',
  });

  const res = await localAwareFetch(url, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body,
  });

  if (!res.ok) {
    const text = await readErrorBody(res);
    const parsed = parseOAuthErrorBody(text);
    throw new AuthFlowError({
      stage: 'authorize',
      statusCode: res.status,
      url,
      body: text,
      errorCode: parsed.errorCode,
      errorDescription: parsed.errorDescription,
      message: parsed.errorDescription || parsed.errorCode || `Authorize failed (${res.status})`,
    });
  }

  const data = (await res.json()) as AuthorizeResponse;

  if (!data.flowId) {
    throw new AuthFlowError({
      stage: 'authorize',
      statusCode: res.status,
      url,
      errorCode: 'missing_flow_id',
      errorDescription: 'Authorize response did not contain a flowId',
      body: JSON.stringify(data),
    });
  }

  const authenticatorId = data.nextStep?.authenticators?.[0]?.authenticatorId;
  if (!authenticatorId) {
    throw new AuthFlowError({
      stage: 'authorize',
      statusCode: res.status,
      url,
      errorCode: 'no_authenticator',
      errorDescription: 'Authorize response did not contain an authenticator',
      body: JSON.stringify(data),
    });
  }

  return { flowId: data.flowId, authenticatorId };
}

async function submitCredentials(
  baseUrl: string,
  flowId: string,
  authenticatorId: string,
  agentId: string,
  agentSecret: string
): Promise<string> {
  const url = `${baseUrl}/oauth2/authn`;
  const payload = {
    flowId,
    selectedAuthenticator: {
      authenticatorId,
      params: {
        username: agentId,
        password: agentSecret,
      },
    },
  };

  const res = await localAwareFetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const text = await readErrorBody(res);
    const parsed = parseOAuthErrorBody(text);
    throw new AuthFlowError({
      stage: 'authn',
      statusCode: res.status,
      url,
      body: text,
      errorCode: parsed.errorCode || (res.status === 401 ? 'invalid_credentials' : undefined),
      errorDescription:
        parsed.errorDescription ||
        (res.status === 401 ? 'Agent ID or secret was rejected by the IAM' : undefined),
      message: parsed.errorDescription || parsed.errorCode || `Authn failed (${res.status})`,
    });
  }

  const data = (await res.json()) as AuthnResponse;

  // Asgardeo can return 200 OK with a FAIL_INCOMPLETE / INCOMPLETE flow status
  // when credentials don't pass. Detect those cases too.
  if (data.flowStatus && data.flowStatus !== 'SUCCESS_COMPLETED') {
    throw new AuthFlowError({
      stage: 'authn',
      statusCode: res.status,
      url,
      body: JSON.stringify(data),
      errorCode: data.error || data.flowStatus,
      errorDescription:
        data.error_description || data.failureReason || `Authn rejected: ${data.flowStatus}`,
    });
  }

  const code = data.authData?.code ?? data.code;
  if (!code) {
    throw new AuthFlowError({
      stage: 'authn',
      statusCode: res.status,
      url,
      body: JSON.stringify(data),
      errorCode: 'no_authorization_code',
      errorDescription: 'Authn succeeded but no authorization code was returned',
    });
  }

  return code;
}

async function exchangeCodeForToken(
  baseUrl: string,
  clientId: string,
  redirectUri: string,
  code: string,
  codeVerifier: string
): Promise<TokenResponse> {
  const url = `${baseUrl}/oauth2/token`;
  const body = new URLSearchParams({
    grant_type: 'authorization_code',
    client_id: clientId,
    code,
    code_verifier: codeVerifier,
    redirect_uri: redirectUri,
  });

  const res = await localAwareFetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body,
  });

  if (!res.ok) {
    const text = await readErrorBody(res);
    const parsed = parseOAuthErrorBody(text);
    throw new AuthFlowError({
      stage: 'token',
      statusCode: res.status,
      url,
      body: text,
      errorCode: parsed.errorCode,
      errorDescription: parsed.errorDescription,
      message: parsed.errorDescription || parsed.errorCode || `Token exchange failed (${res.status})`,
    });
  }

  return res.json() as Promise<TokenResponse>;
}

export async function authenticateAgent(config: AgentAuthConfig): Promise<string> {
  const baseUrl = config.baseUrl.replace(/\/+$/, '');
  const scope = config.scope?.trim() || 'openid';

  const { codeVerifier, codeChallenge } = generatePKCE();
  const { flowId, authenticatorId } = await initiateAuthorize(
    baseUrl,
    config.clientId,
    config.redirectUri,
    scope,
    codeChallenge
  );
  const code = await submitCredentials(baseUrl, flowId, authenticatorId, config.agentId, config.agentSecret);
  const tokenResponse = await exchangeCodeForToken(baseUrl, config.clientId, config.redirectUri, code, codeVerifier);

  if (!tokenResponse.access_token) {
    throw new AuthFlowError({
      stage: 'token',
      url: `${baseUrl}/oauth2/token`,
      errorCode: 'no_access_token',
      errorDescription: 'Token endpoint returned 200 but no access_token field',
      body: JSON.stringify(tokenResponse),
    });
  }

  return tokenResponse.access_token;
}
